### PR TITLE
Remove `as_json` from value and tests

### PR DESCRIPTION
**Pull request by eileencodes**
TL;DR `as_json` is slow. We're trying to replace JSON with Yajl but are hitting problem with the weirdness that are the monkey patches around JSON in AS. We found that some JSON is getting parsed 3 times! and one of the calls was supposed to be [deprecated awhile ago](https://github.com/rails/rails/pull/26933#issuecomment-260185056). Can we remove this without deprecation? If not can we deprecate it in 5.2 even though that's non-standard?

---

Removing `as_json` here revealed that these tests failed because they
return a non-json compliant value of NaNNumber and InfinitNumber.

While we've supported this for awhile one of the issues with this
support is that once we hit `jsonify` all the values _should_ already be
json-compliant values. These tests I removed were added in #26933 - but
in the comments on that PR they were supposed to be deprecated and
removed in the next cycle - but that work never happened. See
https://github.com/rails/rails/pull/26933#issuecomment-260185056

We want to remove this `as_json` here because Rails JSON is very slow.
One of the reasons it's slow is because in some cases we actually loop
over the JSON we're parsing 3 times!

Here are the passes:

1. Call as_json on a tree of objects
  1a. Special objects pass the options down. These are
  Object, Struct, Enumerable, Array, Hash
  1b. This first pass eliminates Structs, Enumerables, and Objects
  Structs replaced with hashes, Enumerables -> Arrays, Objects -> Hash
  1c. New tree contains only: String, Number, Hash, Array, nil, true,
  false
2. Tree is walked again
  2a. Convert Strings to EscapedStrings
  2b. Can retun bad values from as_json (like the infinity and nan)
3. Convert to json

By removing this `as_json` we reduce one of the passes for this case.
The argument here is that the first pass should never have returned a
bad value so we shouldn't have to walk those values a second time - the
value should be `nil` (which it what it will be on pass 3).

[Eileen M. Uchitelle & Aaron Patterson]

cc/ @rafaelfranca @tenderlove @matthewd @jeremy (who else is json knowledgable?)

### CODE REVIEW

**Comment by rafaelfranca** [comment]
It seems removing that code Rails generate invalid JSON, is that right?

> once we hit `jsonify` all the values _should_ already be
> json-compliant values.

I think the documentation of this method states the opposite 

```
           # Recursively calls #as_json to the object to recursively build a
           # fully JSON-ready object.
           #
           # This allows developers to implement #as_json without having to
           # worry about what base types of objects they are allowed to return
           # or having to remember to call #as_json recursively.
```

**Comment by matthewd** [comment]
Is there a distinction to be drawn here between the specific classes that JSON knows (`Integer` & `Float`?), and more general subclasses of `Numeric`, which may well need another round of `as_json` to turn them into something directly serializable?

**Comment by tenderlove** [comment]
@rafaelfranca I think we're doing the same work twice.  `jsonify` is a private method, and we call the method [with the result of `as_json` on the target tree](https://github.com/rails/rails/blob/b802e08273f899d5f3b199f7c6a4f5d514c1b0e1/activesupport/lib/active_support/json/encoding.rb#L34-L36).  The [`as_json` monkey patches](https://github.com/rails/rails/blob/b802e08273f899d5f3b199f7c6a4f5d514c1b0e1/activesupport/lib/active_support/core_ext/object/json.rb) will walk the object tree and *should* return an object tree that is JSON safe.

Note that when I say "JSON safe" I mean "these objects can be represented in JSON" **not** "these objects are escaped and safe to be embedded in HTML".

The `as_json` call [here](https://github.com/rails/rails/blob/b802e08273f899d5f3b199f7c6a4f5d514c1b0e1/activesupport/lib/active_support/json/encoding.rb) will return an object tree that is converted to primitives.  I think the main point of the `jsonify` method is [this line](https://github.com/rails/rails/blob/b802e08273f899d5f3b199f7c6a4f5d514c1b0e1/activesupport/lib/active_support/json/encoding.rb#L92) which deals with escaping certain strings.

Today, we'll actually call `as_json` on everything twice when we try to serialize it.

Here is an example:

```ruby
require 'active_support/all'

class Integer
  alias :old_as_json :as_json

  def as_json(options = nil)
    puts "as_json called on #{self}"
    old_as_json(options)
  end
end

ActiveSupport::JSON.encode([1, 2, 3])
```

If you run this program the output will be:

```
$ be ruby test.rb
as_json called on 1
as_json called on 2
as_json called on 3
as_json called on 1
as_json called on 2
as_json called on 3
```

**Comment by tenderlove** [comment]
I think the main thing about this PR is that we need to change the contract for `as_json` implementations.  Today we support users implementing an `as_json` method that *doesn't* return JSON-safe objects.  We support that by walking the object tree and calling `as_json` on everything twice.  If we insist that the contract is "`as_json` must return JSON safe objects", then we only need to call `as_json` on each object once.

**Comment by rafaelfranca** [comment]
Makes sense to me. Would not this be a "hard to detect" breaking change for applications that doesn't return JSON-safe objects in `as_json`? If you method sometimes doesn't return a JSON-safe object today users will get no error and after this change they will get errors sometimes.

**Comment by tenderlove** [comment]
> Would not this be a "hard to detect" breaking change for applications that doesn't return JSON-safe objects in `as_json`?

Ya, I think it would be hard to detect.  As @eileencodes pointed out, we were *supposed* to deprecate support for this.  In Rails 5.2 we are doing two passes on the object graph, and we could emit a warning or raise an exception.  If we only do one pass on the object graph, we would not detect it.

Maybe we could continue to do two passes in development and test, but only do one pass in production?  Then in dev / test we can raise an exception.

Alternatively, we could provide a "one pass" solution that people opt-in to, add deprecations to the two pass emitter, then change the default emitter in the next version.

**Comment by rafaelfranca** [comment]
> Alternatively, we could provide a "one pass" solution that people opt-in to, add deprecations to the two pass emitter, then change the default emitter in the next version.

I'd go with this one. That way we can use the deprecations to identify cases in production that would start to fail in 6.1

**Comment by tenderlove** [comment]
I have an idea.  We may be able to reduce this to one pass *most* of the time and still support the existing behavior.  If we flip from using the `as_json` monkey patches to using a visitor pattern, then we could detect when a user implements `as_json` and we can verify the results of the `as_json` call.  That would limit the two passes to only the tree subset that has custom `as_json` methods.

I think the trick with that approach is that we need to keep `as_json` at least on `Object` for backwards compatibility, and then the visitor object will have to detect that the `as_json` method is *our* `as_json` method and not a user's `as_json` method.  Doing this check could slow down the single pass emitter, but I'm not sure by how much.

I'll write a quick PoC to give a concrete example of what I mean for implementation.

**Comment by tenderlove** [comment]
Sorry this code is so long:

```ruby
require 'json'

# This is in Rails
class Object
  def as_json
    self
  end
end

module AS
  module JSON
    def self.encode object
      ::JSON.dump JSONTranslator.new.visit object
    end
  end
end

class JSONTranslator
  def visit object
    if object.respond_to?(:as_json) && object.method(:as_json).owner != Object
      return JSONTranslatorWithWarning.new.visit object.as_json
    end

    case object
    when Array
      object.map { |v| visit v }
    when Hash
      Hash[object.map { |v| visit v }]
    when Range
      object.to_a.map { |v| visit v }
    else
      object
    end
  end
end

class JSONTranslatorWithWarning < JSONTranslator
  def visit object
    case object
    when Range
      warn "Ranges not allowed"
    end
    super
  end
end

# This is in User code
class CustomSupported
  def as_json
    "neat"
  end
end

class CustomWarning
  def as_json
    1..2
  end
end

p AS::JSON.encode([1, 2, 3, CustomSupported.new])
p AS::JSON.encode([1, 2, 3, CustomWarning.new])
```

This will only do two passes on *user* objects that implement `as_json`.  I am slightly concerned about the performance of that "respond_to" part in the `visit` method, but I think if we are clever about writing a cache we could eliminate that overhead.

**Comment by rafaelfranca** [comment]
That looks good to me. Even that `respond_to?` add  overhead that is only temporary until we can remove the deprecated code.

**Comment by chancancode** [comment]
Hi!

The intent of the API to support use cases like this:

```rb
class Blog
  # ...

  def as_json
    {
      id: id,
      name: "#{first_name} #{last_name}",
      posts: posts
    }
  end
end
```

...as opposed to something like...

```rb
class Blog
  # ...

  def as_json(options)
    {
      "id" => id,
      "name" => "#{first_name} #{last_name}",
      "posts" => posts.try(:map){ |p| p.as_json(options) }
    }
  end
end
```

At the time we believe the latter is both too unergonomic and also very difficult for the average developer to get right (specifically, putting the responsibility of the recursion on them, peeling off and passing the nested options correctly, etc). It is also a composition hazard, since any of the intermediate layers could forget to do it, and we will get inconsistent serialization at different layers for common types like date.

So the compromise is that, the contract for `as_json` is a "pre-serialization" hook that allows you to do as much or as little transformation you care about, and Rails will "take care of the rest". Specifically, if you return a "JSON primitive" (something that has "obvious" mapping to JSON types), we won't touch it. Otherwise, you should be generally happy Rails doing whatever* serialization it wants with the rest, including throwing an error (* within the reasonable intra-version stability expectations). If you are not happy with the result for some reason, that means you care about that part of the tree after all and you should have transformed it in `as_json`.

The consequences are:

1) I think it would be totally reasonable to throw when you give us `Infinity`
2) It is generally okay to expand the set of things we don't handle, as long as we carefully considered the amount of breakage and ergonomics impact and break it at the appropriate version boundary (which varies given that specific change)
3) Given the API contract (I didn't really come up with it, I just rationalized what was then a random set of ad-hoc things into something slightly more coherent), some amount of re-walking is hard to avoid, especially since we now delegate to the native `JSON` encoder for the actual string-writing (otherwise some of the work could happen in that walk)

**Comment by chancancode** [comment]
I think the last point is important and explains why we find ourselves in this situation. It used to be that we do the encoding ourselves, so certain things are easier to support "efficiently" – in terms of number-of-passes, even though on the whole it was much slower to do the actual serialization in pure ruby compared to C. So, in order to fix the bigger performance problem, we tried to replace the serialization pass with just the `JSON` gem, but in order to maintain compatibility we have to do a pre-walk to normalize differences between the `JSON` gem and the old behavior, because we lost the ability to do those things "on the fly". The problem then is that we can't simply call `as_json` and pass that to the encoder, because we can't trust the developers, and `as_json` was really not meant/designed for that job anyway (the first example is the norm/main use case). So we ended up traded the encoding performance with an extra pass. The gain in encoding performance easily paid for the extra pass, but that left us at where we are today.

**Comment by chancancode** [comment]
I think removing `Infinity` and `NaN` support is a no brainer though. It's definitely not obvious that `null` is the right serialization there (similar to [this](https://github.com/rails/rails/issues/25017#issuecomment-238186031)) so I think it should be up to the developers to handle anyway for correctness. If we get some performance wins out of that too, then doubly great.

**Comment by chancancode** [comment]
> Specifically, if you return a "JSON primitive" (something that has "obvious" mapping to JSON types), we won't touch it. Otherwise, you should be generally happy Rails doing whatever* serialization it wants with the rest, including throwing an error (* within the reasonable intra-version stability expectations). If you are not happy with the result for some reason, that means you care about that part of the tree after all and you should have transformed it in `as_json`.

To be clear, this is a very generic description. So generic that even if we stopped doing the second pass (so that "whatever Rails does" === "whatever JSON does") it would still be consistent with that contract (though obviously it would be a major-version-breaking-change). I just don't personally think that would get us to a good place in terms of ergonomics and ecosystem-wide coherence. (Imagine filing bugs to all the gems that implements `as_json`.)

I actually think that it is generally good for people to do as little as possible in their own `as_json`, so that we could own the tree-walk for the most part and do whatever optimizations we can. (Re-implementing a custom Rails JSON encoder natively is still a thing I want to try to do at some point.) The moment we push the responsibility mostly back to the developers in their own `as_json`, I think we will lose a lot of control over that which would be bad in the long term, both in from the performance and functionality perspective.

**Comment by chancancode** [comment]
> I have an idea. We may be able to reduce this to one pass most of the time and still support the existing behavior. If we flip from using the `as_json` monkey patches to using a visitor pattern, then we could detect when a user implements `as_json` and we can verify the results of the `as_json` call. That would limit the two passes to only the tree subset that has custom `as_json` methods.

I think we can certainly try to defer the work and coalesce the passes, however there are some complexities involved with that. Specifically something like this:

```ruby
class Blog
  # ...

  def as_json(*)
    {
      id: id,
      name: "#{first_name} #{last_name}",
      posts: posts.as_json(only: ...)
    }
  end
end
```

We either make `Hash#as_json` return some kind of `LazyAsJsonResultWithTheOptionsClosedInsideIt` object, or at least make it a no-op (i.e. return `self`) for the common case of no options (although imo you should just not call `as_json` recursively there if you don't need to do anything custom).

In any case, we can probably get away with removing [this](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/json.rb#L172) and try to move/coalesce it into the encoder pass. IIRC, the main reason this is there is because people treat `as_json` as a Public API and have some expectations about its return value (e.g. the keys would be stringified after you call it). However I personally don't really think that's a thing. I have repeatedly said that only Rails or implementations of `as_json` should call `as_json`, and you shouldn't have any expectations about their return value other than "when fed into the encoding process the right thing will happen". In other words, it is a hook for Rails to call and for people to override, not a general-purpose utility method.

**Comment by tenderlove** [comment]
@chancancode I think we can do a single pass visitor without returning a special object from `as_json`.  Objects that implement `as_json` will have a performance penalty, but objects that don't won't.  Moreover, I think the performance penalty will be equal to the current two pass penalty today.

**Comment by chancancode** [comment]
@tenderlove I think we are mostly talking about the same thing, but instead of trying to detect/special-case the `as_json` call, we can just call it all the time, but have it do nothing most of the time:

```rb
require 'json'

# This is in Rails
class Object
  # exactly same code as today
  def as_json(options = nil)
    if respond_to?(:to_hash)
      to_hash.as_json(options)
    else
      instance_values.as_json(options)
    end
  end
end

class Array
  def as_json(options = nil)
    if options
      # slow path: same as today
      # do we really need the dup???
      map { |v| v.as_json(options.dup) }
    else
      # fast path: new
      self
    end
  end
end

class Hash
  def as_json(options = nil)
    if options
      # slow path: exactly the same as today
      subset = if attrs = options[:only]
        slice(*Array(attrs))
      elsif attrs = options[:except]
        except(*Array(attrs))
      else
        self
      end

      # do we really need the dup???
      Hash[subset.map { |k, v| [k, v.as_json(options.dup)] }]
    else
      # fast path: new
      self
    end
  end
end

module AS
  module JSON
    def self.encode(object, options = nil)
      if option
        # slow path: extra pass here
        object = object.as_json(options)
      end

      ::JSON.generate case object
        when Array
          # this does the actual walk, normally ("fast path" above) this would be the only walk other 
          object.map { |v| v.as_json }
        when Hash
          Hash[object.map { |k, v| [k.as_json, v.as_json] }]
        else
          object.as_json
        end
      end
  end
end
```

Something like this. In most cases `as_json` should be called only once.

**Comment by chancancode** [comment]
(IMO the `as_json` options thing itself is pretty inconsistent anyway. If we think we can deprecate and remove _that_ and move people to using serializers/jbuilder, it would remove a lot of complexity.)

**Comment by tenderlove** [comment]
@chancancode I don't think that eliminates the two passes.  You'll still have to do a second pass in order to support [escaped strings](https://github.com/rails/rails/blob/b802e08273f899d5f3b199f7c6a4f5d514c1b0e1/activesupport/lib/active_support/json/encoding.rb#L92).

**Comment by chancancode** [comment]
Sure, I omitted some details to match your example. That would take place in the same place where `case Hash` and `case Array` where the actual work happen. My overall point is calling `as_json` is not intrinsically slow, because most of the time it shouldn't have to do anything at all – as long as we are okay breaking people's expectation about the *return value* of `as_json`.

I'm not exactly sure how we are counting the passes (specifically, whether the actual encoding pass inside `JSON.generate` count as one), but I *think* we are talking about the same total number of passes.

**Comment by tenderlove** [comment]
@chancancode check my example here: https://github.com/rails/rails/pull/34578#issuecomment-444591846

It calls `as_json` on all integers twice, and that's before passing it to the JSON gem for encoding.  The total passes on the object graph is 3 if you include the JSON gem.

**Comment by chancancode** [comment]
Yeah, I understand that.

In the case of integers (numeric), the second pass was introduced in the patch @eileencodes linked at the top to support `Infinity` and `NaN`. I personally think we can break that, which would remove the second pass in that case. In any case, numbers are probably *not* the high-order bit since they are by definition shallow/flat and the work in their `as_json` is cheap (we should still avoid it, and I think we can).

On the other hand, `Hash` and `Array` are deep-walked at least twice which IMO is more problematic than numbers. The first pass is at the top of `encode` as you said, which calls `#as_json` on `Hash`/`Array`, which does their own deep walk, and then later a second pass in `jsonify`. I am suggesting to remove the pass inside `#as_json` when encoding without options (i.e. `return self`), but keep the pass inside `encode` as the primary pass that always happen. This should be okay if 1) we don't care about providing any guarantees about what `as_json` return (already true IMO) and 2) we don't allow overriding `as_json` on `Hash` and `Array` subclasses working reliably (maybe more controversial).
